theano/gpuarray/opt.py

```diff
@@ -210,6 +210,52 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(Optimizer):
+    """
+    Transfer the graph as a whole to GPU instead of replacing nodes
+    """
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        move_to_GPU = True
+
+        # Building a new graph
+        for i in fgraph.inputs:
+            mapping[i] = GpuFromHost(None)(i)
+
+        for node in fgraph.toposort():
+
+            # The Extra condition
+            if node.inputs is node.outputs:
+                move_to_GPU = False
+
+            # Oplifter's condition
+            # Will return a list of OP
+            # If None, means can't be moved.
+
+            new_ops = local_gpuaalloc(node, None)
+
+            if new_ops is None:
+                move_to_GPU = False
+
+            if not isinstance(new_ops[0], node.op):
+                move_to_GPU = False
+                continue
+
+            if move_to_GPU:
+                for old_o, new_o in zip(node.outputs, new_ops):
+                    mapping[old_o] = new_o
+            else:
+                for o in node.outputs:
+                    mapping[o] = o
+
```

Comment:
@nouiz I tried that by applying this logic, using `orderings` method , assuming that it will return all the nodes that occur prior to outputs.

```
        for o in fgraph.outputs:
            current_dependency = set(ordering[o])
            shared_node = current_dependency.interesection(prev_dependency)
            if not shared_node:
                original_node.append(o)
                new_o = mapping[o]
                if new_o.type != o.type:
                    assert isinstance(o.type, tensor.TensorType)
                    assert isinstance(new_o.type, GpuArrayType)
                    new_o = host_from_gpu(new_o)
                rep_node.append(new_o)
            prev_dependency = current_dependency
        for k,l in zip(prev_dependency, current_dependency):
            fgraph.replace_validate(k, l)
```
This is giving me a KeyError.